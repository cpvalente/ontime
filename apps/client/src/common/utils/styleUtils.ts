import Color from 'color';

type ColourCombination = {
  backgroundColor: string;
  color: string;
};

/**
 * @description Selects text colour to maintain accessible contrast
 * @param bgColour
 * @return {{backgroundColor, color: string}}
 */
export const getAccessibleColour = (bgColour?: string): ColourCombination => {
  if (bgColour) {
    try {
      const originalColour = Color(bgColour);
      const backgroundColorMix = originalColour.alpha(1).mix(Color('#1a1a1a'), 1 - originalColour.alpha());
      const textColor = backgroundColorMix.isLight() ? 'black' : '#fffffa';
      return { backgroundColor: backgroundColorMix.hexa(), color: textColor };
    } catch (_error) {
      /* we do not handle errors here */
    }
  }
  return { backgroundColor: '#1a1a1a', color: '#fffffa' };
};

/**
 * @description Creates a list of classnames from array of css module conditions
 * @param classNames - css modules objects
 */
export const cx = (classNames: any[]) => classNames.filter(Boolean).join(' ');

export const enDash = '–';

export const timerPlaceholder = '––:––:––';
export const timerPlaceholderMin = '––:––';
