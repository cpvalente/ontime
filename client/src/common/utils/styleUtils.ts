import Color from 'color';

type ColourCombination = {
  backgroundColor: string;
  color: string;
}

/**
 * @description Selects text colour to maintain accessible contrast
 * @param bgColour
 * @return {{backgroundColor, color: string}}
 */
export const getAccessibleColour = (bgColour: string): ColourCombination => {
  if (bgColour) {
    try {
      const textColor = Color(bgColour).isLight() ? 'black' : '#fffffa';
      return { backgroundColor: bgColour, color: textColor };
    } catch (error) {
      console.log(`Unable to parse colour: ${bgColour}`);
    }
  }
  return { backgroundColor: '#000', color: "#fffffa" };
};
