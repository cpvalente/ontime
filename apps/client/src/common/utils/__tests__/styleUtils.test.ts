import { cx, getAccessibleColour } from '../styleUtils';

import style from './styleUtils.module.scss';

describe('cx()', () => {
  it('merges styles', () => {
    const merged = cx([style.test, style.another]);
    expect(merged).toMatchSnapshot();
  });
  it('ignores falsy values', () => {
    const falsyStuff = false;
    const merged = cx([undefined, false, 0, null, falsyStuff ? style.test : null]);
    expect(merged).toMatchSnapshot();
  });
});

describe('getAccessibleColour()', () => {
  it('handles named colours', () => {
    const colour = 'red';
    const { backgroundColor, color } = getAccessibleColour(colour);
    expect(backgroundColor).toBe('#FF0000FF');
    expect(color).toBe('#fffffa');
  });
  it('handles hex colours', () => {
    const colour = '#0F0';
    const { backgroundColor, color } = getAccessibleColour(colour);
    expect(backgroundColor).toBe('#00FF00FF');
    expect(color).toBe('black');
  });
  it('handles transparens', () => {
    const colour = '#0F08';
    const { backgroundColor, color } = getAccessibleColour(colour);
    expect(backgroundColor).toBe('#0C940CFF');
    expect(color).toBe('#fffffa');
  });
});
