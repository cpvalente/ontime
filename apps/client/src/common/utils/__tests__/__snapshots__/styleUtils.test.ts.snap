// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cx() > ignores falsy values 1`] = `""`;

exports[`cx() > merges styles 1`] = `"_test_98a1e0 _another_98a1e0"`;
